llama-index-integrations/vector_stores/llama-index-vector-stores-milvus/llama_index/vector_stores/milvus/base.py

```diff
@@ -284,12 +284,64 @@ def __init__(
             if dim is None:
                 raise ValueError("Dim argument required for collection creation.")
             if self.enable_sparse is False:
-                schema = self._create_schema()
-                self._milvusclient.create_collection(
-                    collection_name=collection_name,
-                    schema=schema,
-                    consistency_level=consistency_level,
-                )
+                # Check if custom index should be created
+                if (
+                    index_config is not None
+                    and self.index_management is not IndexManagement.NO_VALIDATION
+                ):
+                    try:
+                        # Prepare index
+                        index_params = self.client.prepare_index_params()
+                        index_type = index_config["index_type"]
+                        index_params.add_index(
+                            field_name=embedding_field,
+                            index_type=index_type,
+                            metric_type=self.similarity_metric,
+                        )
+
+                        # Create a schema according to LlamaIndex Schema.
+                        schema = self._create_schema()
+                        schema.verify()
+
+                        # Using private method exposed by pymilvus client, in order to avoid creating indexes twice
+                        # Reason: create_collection in pymilvus only checks schema and ignores index_config setup
+                        # https://github.com/milvus-io/pymilvus/issues/2265
+                        self.client._create_collection_with_schema(
+                            collection_name=collection_name,
+                            schema=schema,
+                            index_params=index_params,
+                            dimemsion=dim,
+                            primary_field=MILVUS_ID_FIELD,
+                            vector_field=embedding_field,
+                            id_type="string",
+                            max_length=65_535,
+                            consistency_level=consistency_level,
+                        )
+                        self._collection = Collection(
+                            collection_name, using=self._milvusclient._using
+                        )
+                    except Exception as e:
+                        logger.error("Error creating collection with index_config")
+                        raise NotImplementedError(
+                            "Error creating collection with index_config"
+                        ) from e
+                else:
+                    self._milvusclient.create_collection(
```

Comment:
Here we should also call `schema = self._create_schema()` and pass it to create_collection like above. Else the user-defined fields (`scalar_field_names`, `scalar_field_types`) will be ignored if the user does not provide the index_config parameter.

Reply:
I added this, but when I added a test, it also broke the code with the ambiguous index error again

Going to leave this out, since my test will fail if I add schema here

Reply:
Since the tests pass, I'd rather get a working PR published. Feel free to open a PR to further address this though